### ISSUE TITLE
CI3 and PHP8.1(ctype_digit(): Argument of type int will be interpreted as string in the future) 

### ISSUE THREAD

**Issue by Obata2024** [bcit-ci/CodeIgniter#6285]
When trying to use PostgreSQL under the given conditions, the following error occurs:

> A PHP Error was encountered
> Severity: 8192
> 
> Message: ctype_digit(): Argument of type int will be interpreted as string in the future
> 
> Filename: postgre/postgre_driver.php
> 
> Line Number: 98

Should I modify the postgre/postgre_driver.php file to resolve this? If it’s acceptable, I can submit a PR.

Also, has anyone resolved the same issue using an override?
I attempted to create a custom class under the application/core directory to override it, but this error still persists.

**Comment by jamieburchell** [comment]
Looks to me that the ctype part of that if statement should just be removed. As far as I can see so far none of the other drivers check anything other than that the value isn't empty before using it.

**Comment by daveherman71** [comment]
Removing the check might have unintended consequences so a better approach would be to simply cast the variable as a string which is done pretty much everywhere else in the framework.

Change line 98 in the file to read:
```
if ( ! empty($this->port) && ctype_digit((string) $this->port))
```

**Comment by jamieburchell** [comment]
> Removing the check might have unintended consequences so a better approach would be to simply cast the variable as a string which is done pretty much everywhere else in the framework.

It's inconsistently checked [in](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/database/drivers/cubrid/cubrid_driver.php#L108) [each](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/database/drivers/mssql/mssql_driver.php#L96) [database](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/database/drivers/mysqli/mysqli_driver.php#L138) [driver](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/database/drivers/oci8/oci8_driver.php#L180). Most just check if the port is not empty. In this case, it's just to find out if the port number is not empty. If casting to a string, this should also be applied elsewhere such as [here](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/database/drivers/oci8/oci8_driver.php#L180) and [here](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/database/drivers/oci8/oci8_driver.php#L190).

**Comment by daveherman71** [comment]
Correct. Also in [DB_driver.php](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/database/DB_driver.php#L1351), [DB_query_builder.php](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/database/DB_query_builder.php#L2785), [Session.php](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/libraries/Session/Session.php#L393), [Form_validation.php](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/libraries/Form_validation.php#L1215) and [Image_lib.php](https://github.com/bcit-ci/CodeIgniter/blob/3658d731eaabe6117298a105ffb5b9dd59e190ce/system/libraries/Image_lib.php#L634)

**Comment by jamieburchell** [comment]
The thing is, in some of these cases the variable is expected to be a string and not an int, for example in the `valid_url` function. The lack of type hinting on the method parameters is biting us here.

Either way, the simplest fix here is casting. I'll open a PR later and add it to my fork.
